Add pseudo potential

ndaelman-hu opened the pull request:
Some users have been requesting pseudopotential support. This was available in the runschema and should therefore be ported anyhow. The schema was further extended to include additional metadata with codes like CASTEP and QE in mind.
It also leverages other schema-generic pieces for relativity and DFT annotation.

ndaelman-hu committed:
Add Pseudopotential class - direct migration from runschema

Ported Pseudopotential class from nomad-schema-plugin-run to numerical_settings.py.
Provides high-level metadata for pseudopotentials: type, cutoff, XC functional, and angular momentum parameters.

ndaelman-hu committed:
Clarify distinction between Pseudopotential and EffectiveCorePotential

Updated docstrings to explain that these classes serve non-overlapping domains:
- Pseudopotential: metadata for plane-wave DFT codes (VASP, QE, CASTEP)
- EffectiveCorePotential: analytical ECPs for Gaussian basis set codes

Key clarifications:
- PAW/ultrasoft PPs cannot be represented as ECPs
- Plane-wave PPs store numerical tables, not Gaussian expansions
- Full PP data is external (POTCAR, UPF files), only metadata stored

ndaelman-hu committed:
Add pseudopotential reference to AtomsState

Integrate Pseudopotential with AtomsState following combo A+C pattern:
- Pseudopotential instances defined in numerical_settings (standalone)
- AtomsState references them via typed reference using SectionProxy

This allows parsers to populate pseudopotentials in numerical_settings and link
them to specific atomic species, maintaining clear separation between structural
information (atoms) and computational settings (pseudopotentials).

ndaelman-hu committed:
Add r_core, pseudization_scheme, and relativistic to Pseudopotential

Extended Pseudopotential metadata based on CASTEP documentation analysis:

- `r_core`: Core/augmentation radius with nuanced description for NC/US/PAW
- `pseudization_scheme`: Method used to generate smooth PP (TM, Polynomial, Bessel, Extra-Soft, unavailable)
- `relativistic`: Level of relativistic treatment (non/scalar/full-relativistic, unavailable)

These additions provide important metadata about PP construction and applicability.
Includes 'unavailable' option following schema pattern for unspecified values.

ndaelman-hu committed:
Use RelativityModel for pseudopotential relativistic treatment

Replaced simple relativistic enum with SubSection reference to RelativityModel.
This provides consistent schema across calculation and PP generation, allowing
specification of both level (scalar, two-component, etc.) and approximation
(DKH, ZORA, X2C, etc.) used during pseudopotential generation.

Benefits:
- Consistent with existing RelativityModel in model_method
- More detailed: captures both level and specific approximation
- Clear distinction: PP generation (frozen) vs calculation (runtime)

ndaelman-hu committed:
Add n_valence_electrons and reference_configuration to Pseudopotential

Added two related quantities for documenting pseudopotential valence treatment:

- `n_valence_electrons`: Number of electrons explicitly treated as valence (VASP: ZVAL,
  QE: z_valence, CASTEP: ionic charge). Universal across all plane-wave codes.

- `reference_configuration`: Electronic configuration string used during PP generation
  (e.g., "3p6 3d7 4s1"). May omit deeper semi-core levels that are counted in n_valence_electrons.

Both describe valence electrons (explicitly treated), not core electrons (frozen in PP).
The quantities are related but configuration string may be abbreviated.

ndaelman-hu committed:
Clarify cutoff description for multi-level recommendations

Updated cutoff description to specify that when multiple recommendations exist
(e.g., CASTEP coarse/medium/fine, VASP ENMIN/ENMAX), parsers should store the
most representative or medium value. Codes with sophisticated cutoff systems
can extend the schema at parser-specific level.

Keeps core schema simple while providing clear guidance for parser implementations.

ndaelman-hu committed:
Add `gw_optimized` flag to Pseudopotential

GW-optimized pseudopotentials (e.g., VASP _GW suffix) are validated for
scattering properties far above the Fermi level and typically include more
semi-core states for accurate excited-state calculations (GW, BSE, optical).
They remain valid for standard DFT but are computationally more expensive.

ndaelman-hu committed:
Use XCFunctional SubSection for pseudopotential XC functional

Replace simple string array `xc_functional_name` with structured
`XCFunctional` subsection to align with DFT schema. This allows parsers
to use the same LibXC canonicalization (functional_key expansion) and
provides consistent XC metadata across schema.

Fix circular import between numerical_settings.py and model_method.py by
using SectionProxy for both XCFunctional and RelativityModel instead of
direct imports.

ndaelman-hu committed:
Apply ruff formatting

ndaelman-hu commented:
@EBB2675 I used the sections `RelativityModel` and `XCFunctional`, which you recently added.
Pls also remark on how you see their wider spread usage, as well as import handling.

ndaelman-hu commented:
@JFRudzinski if you want, I can test this schema out right-away in a few parsers. The first candidates would be: VASP, FHIaims, CASTEP, and QE. Aside from the obvious parsing extension, this has a 2-pronged advantage:

- some codes distinguish multiple cutoffs for different levels of precision. I already indicated that this should be extended on their side. This could be the actual test.
- I might also get feedback from the CASTEP devs. This isn't necessarily contingent on us extending the parser, but it would direct their focus.

EBB2675 commented on src/nomad_simulations/schema_packages/numerical_settings.py:
```diff
@@ -1032,6 +1032,170 @@ class FrozenCore(NumericalSettings):
+    )
+
+    type = Quantity(
+        type=MEnum('US V', 'US MBK', 'PAW'),
```
The description table lists 'US', 'PAW', 'V', 'MBK', and enum is `MEnum('US V', 'US MBK', 'PAW')`

This mismatch might come across as confusing. Right now it’s not clear whether:

- `type` is meant as combination of ultrasoft + builder (US V, US MBK), or
-	you conceptually want a primary type (PAW, US, NC) and a separate “family/builder” classification (V, MBK, …).

I’d at least align the table with the actual enum values or explicitly say “'US V' = Vanderbilt ultrasoft; 'US MBK' = Morrison-Bylander-Kleinman ultrasoft”.

ndaelman-hu replied:
Good catch! Typically people care more so about the primary type, not that much the numerical / builder representation.
Adding the latter would also require some added verification logic. Better to V and MDK out.

On that note, I also re-evaluated NC, as it is sometimes combined with PAW. In particular when PAW pseudopotentials are used for GW.

EBB2675 commented on src/nomad_simulations/schema_packages/numerical_settings.py:
```diff
@@ -1032,6 +1032,170 @@ class FrozenCore(NumericalSettings):
+
+    Pseudopotentials approximate the potential of core electrons and the nucleus, enabling
+    efficient treatment of valence electrons in codes like VASP, Quantum ESPRESSO, and CASTEP.
+    Common types include PAW (Projector Augmented Wave), ultrasoft (US), and norm-conserving (NC).
```
```suggestion
    Common types include PAW (Projector Augmented Wave), ultrasoft (US), and norm-conserving (NC).
    Norm-conserving pseudopotentials are represented by setting `norm_conserving = True` and choosing the appropriate value in `type`.
```

EBB2675 replied:
The docstring talks about “norm-conserving (NC), ultrasoft (US), and PAW” as three types. 

In the schema, norm-conservation is encoded only via the `norm_conserving: bool` flag, while the type enum distinguishes the representation and does not contain an explicit NC value. It might help to add one sentence above that makes this explicit so future contributors don’t expect “NC” to appear as a separate entry in the `type` enumeration.

EBB2675 commented on src/nomad_simulations/schema_packages/numerical_settings.py:
```diff
@@ -1032,6 +1032,170 @@ class FrozenCore(NumericalSettings):
+        type=np.int32,
+        shape=[],
+        description="""
+        Maximum magnetic momentum of the pseudopotential projectors.
```
```suggestion
        Maximum magnetic quantum number of the pseudopotential projectors.
```

EBB2675 reviewed (approved):
looks great, and also the usage of `RelativityModel` looks perfect 👍 

i just added a couple of super minor docstring clarifications

ndaelman-hu committed:
Update src/nomad_simulations/schema_packages/numerical_settings.py

reviewer: clarify how NC PP are represented

Co-authored-by: Esma B. Boydas <80156958+EBB2675@users.noreply.github.com>

ndaelman-hu committed:
Update src/nomad_simulations/schema_packages/numerical_settings.py

reviewer: clarify language for quantum numbers

Co-authored-by: Esma B. Boydas <80156958+EBB2675@users.noreply.github.com>

JFRudzinski reviewed (commented):
It looks good...

I am just wondering a bit about the categorization of pseudo potentials as numerical settings vs. Model method.

I would have thought the identity of the pseudo potential should live as a model method, as they are used to build the Hamilton components. 

Things like cutoff and such, ie, how you evaluate the potentials, I would agree are numerical settings

There is a clear analogy with my force field schema, please check and then we can discuss if it's useful to align these better

ndaelman-hu committed:
Fix Pseudopotential type enum: replace 'US V'/'US MBK'/'PAW' with 'NC'/'US'/'PAW'/'NC-PAW'/'NC-PAW-GW'

ndaelman-hu commented:
> It looks good...
> 
> I am just wondering a bit about the categorization of pseudo potentials as numerical settings vs. Model method.
> 
> I would have thought the identity of the pseudo potential should live as a model method, as they are used to build the Hamilton components.
> 
> Things like cutoff and such, ie, how you evaluate the potentials, I would agree are numerical settings
> 
> There is a clear analogy with my force field schema, please check and then we can discuss if it's useful to align these better

We discussed and agreed that the PP can stay in `numerical_settings.py`.

ndaelman-hu committed:
Add comprehensive pseudopotential documentation: enhanced in-code docstrings and cross-code terminology in package docs

ndaelman-hu committed:
Maximize lazy loading to prevent circular imports

ndaelman-hu committed:
Merge branch 'add_pseudo_potential' of github.com:nomad-coe/nomad-simulations into add_pseudo_potential

ndaelman-hu committed:
Clean pydocs structure

ndaelman-hu committed:
Apply ruff

ndaelman-hu commented on src/nomad_simulations/schema_packages/numerical_settings.py:
```diff
@@ -1032,6 +1032,197 @@ class FrozenCore(NumericalSettings):
+    """
+    Section containing high-level metadata (type, cutoff energy, XC functional) that identifies which pseudopotential was used.
+    Pseudopotentials approximate the potential of core electrons and the nucleus, enabling
+    efficient treatment of valence electrons in codes like VASP, Quantum ESPRESSO, and CASTEP.
```
```suggestion
    efficient treatment of valence electrons in plane-wave codes.
```

ndaelman-hu commented on src/nomad_simulations/schema_packages/numerical_settings.py:
```diff
@@ -1032,6 +1032,197 @@ class FrozenCore(NumericalSettings):
+    projectors at higher energies to accurately describe quasiparticle states far above the Fermi
+    level. Standard PAW and US pseudopotentials systematically underestimate scattering into
+    high-energy unoccupied states, which is critical for GW many-body perturbation theory.
+
```
should go under `type` description

ndaelman-hu replied:
applied

ndaelman-hu commented on src/nomad_simulations/schema_packages/numerical_settings.py:
```diff
@@ -1032,6 +1032,197 @@ class FrozenCore(NumericalSettings):
+        """,
+    )
+
+    norm_conserving = Quantity(
```
```suggestion
    is_norm_conserving = Quantity(
```

ndaelman-hu commented on src/nomad_simulations/schema_packages/numerical_settings.py:
```diff
@@ -1032,6 +1032,197 @@ class FrozenCore(NumericalSettings):
+        when in doubt. Codes with sophisticated multi-level cutoff systems should extend this at
+        the parser-specific schema level.
+        """,
+    )
```
add `cutoff_target` to clarify the reference?

ndaelman-hu replied:
applied

ndaelman-hu commented on .gitignore:
@JFRudzinski are you fine with us adding the default AI folders to the gitignore?

JFRudzinski replied:
yes

ndaelman-hu commented on src/nomad_simulations/schema_packages/basis_set.py:
```diff
@@ -333,9 +333,20 @@ def __init__(self, *args, **kwargs):
+
+    Effective Core Potentials (ECPs) replace core electrons with analytical potentials
+    expressed as sums of Gaussian functions. This representation is used in quantum chemistry
+    codes (Gaussian, ORCA, Molpro, Q-Chem, NWChem) to reduce computational cost while
```
```suggestion
    codes to reduce computational cost while
```

ndaelman-hu commented on src/nomad_simulations/schema_packages/atoms_state.py:
Note: this design appears semantically most desirable, but should first be tested in at least 1 parser

ndaelman-hu replied:
being tested on VASP: https://github.com/FAIRmat-NFDI/nomad-parser-plugins-simulation/pull/129

ndaelman-hu reviewed (commented):
A few of my own thoughts. Will be implemented, then this PR is ready for merging.

ndaelman-hu committed:
Fix RelativityModel SectionProxy resolution

ndaelman-hu committed:
Add normalize() method to Pseudopotential to expand XC functionals

ndaelman-hu committed:
Apply suggestions from code review

- Remove specific callouts to code
- Rename `norm_conserving` to `is_norm_conserving` to reflect `bool` type

ndaelman-hu committed:
Move type classification docs from class to type quantity description

ndaelman-hu committed:
Add cutoff_target field to preserve code-native cutoff terminology

JFRudzinski reviewed (commented):
At a glance it looks good for me :+1:  Someone more of an expert should review in more detail though

ndaelman-hu committed:
Rename  to  to reflect boolean nature

EBB2675 commented on src/nomad_simulations/schema_packages/numerical_settings.py:
```diff
@@ -1032,6 +1032,215 @@ class FrozenCore(NumericalSettings):
+        """,
+    )
+
+    cutoff_target = Quantity(
```
These code-specific strings for plane-wave energy cutoffs are not scientific concepts, but naming conventions in specific codes. Users of PP libraries do not interpret them scientifically, so they are not really necessary for interoperability. Parser authors can resolve these labels internally and store a clean, consistent value in cutoff.

(If in the future a PP format appears that includes multiple scientifically distinct cutoff values, then a slightly different design would be more suitable, for example:

`cutoffs = SubSection(PPCutoff, repeats=True)` )

In particular:

- Does **ENMAX** mean something different than “recommended cutoff”? → No

- Does **ecutwfc** string have a different unit or meaning? → No

- Does knowing the original label help downstream analyses? → No


I am flexible on this, but if we move away from a code-agnostic design, I would need to quickly revisit my quantum-chemistry related implementations and add similar code-specific conventions wherever they are important, similar to these cutoff indicating strings 👍

ndaelman-hu replied:
I quite like your `cutoffs = SubSection(PPCutoff, repeats=True)` suggestion. I guess I still need a representative cutoff for querying: people will definitely want to search pseudopotentials by element and cutoff, those are its main parameters. Or we use nested queries? I could try that.

Either way,  it's not that much a question of "if". Codes like VASP and CP2K provide cutoff ranges or multiple (precision-dependent) cutoffs, respectively, instead of a single representative cutoff. The question I was trying to answer isn't really "Does ENMAX mean something different than “recommended cutoff”? → No", but rather "Which cutoff does NOMAD consider representative: ENMIN or ENMAX?". I guess you did pick up on that to some extent. 

I do not have a much of view on how the choice of `cutoff_target` or `PPCutoff` affects the rest of the schema, though. I hadn't thought about that scope yet, thx for bringing it up.

EBB2675 replied:
cant they query by code name VASP etc instead of cutoff_target?

EBB2675 replied:
i have also checked my previous VASP simulations from my phd, and there were also certain cases in which i didnt resort to any ENMAX or ENMIN, but choose another cutoff

EBB2675 replied:
from my understanding, these recommended cutoffs are provided to guide the user, but at the end, isnt the calculation done by what is set in `ENCUT` ?

if the user does not set `ENCUT` explicitly in the `INCAR`, it indeed takes a default value from POTCAR's largest ENMAX, but still the OUTCAR prints the ENCUT, which is the final cutoff used in any scenario

ndaelman-hu replied:
This has been discussed in a different thread. We clarified that this PR aims to describe PP metadata, not the general basis set. Am resolving this thread.

EBB2675 reviewed (changes requested):
I’d gently lean towards dropping `cutoff_target`: labels like ENMAX/ecutwfc are code-specific rather than scientific, and parsers can already map them internally to a single, clean cutoff value. But I am flexible if others see a strong use case.

if we go for this kind of design, I can have a quick PR on QC schema support, to include analogous, community-known, code-specific conventions as well 👍

EBB2675 commented:
To keep the schema code-agnostic while preserving queryability, we can model cutoffs as *typed physical concepts* rather than code-native labels like `ENMAX` or `ecutwfc`.

### a proposed schema shape, iteration 1 by EBB:

```python
class PPCutoff(ArchiveSection):
    # What physical expansion this cutoff controls
    cutoff_kind = Quantity(
        type=MEnum(
            "wavefunction",     # e.g. QE ecutwfc, VASP ENCUT
            "charge_density",   # e.g. QE ecutrho (often ~4x for NC, higher for US)
            "augmentation",     # PAW augmentation / projector-related cutoffs (if applicable)
            "response",         # response-function basis cutoffs (e.g. ENCUTGW-like)
            "other",
        )
    )

    # What the value means in context
    cutoff_role = Quantity(
        type=MEnum(
            "used",             # actually used in the run (runtime value)
            "recommended",      # vendor/library recommended default
            "recommended_min",  # lower bound recommendation
            "recommended_max",  # upper bound recommendation
            "standard",         # library "standard" recommendation (SSSP-like)
            "stringent",        # library "stringent" recommendation
            "soft",             # explicitly soft setting
            "hard",             # explicitly hard setting
        )
    )

    value = Quantity(type=np.float64, unit="joule")
```

And in `Pseudopotential`:

```python
cutoffs = SubSection(sub_section=PPCutoff.m_def, repeats=True)
```

### here i include some example code-to-semantics mappings

- **VASP `ENMAX`**  
  maps to: `cutoff_kind="wavefunction"`, `cutoff_role="recommended"`  
  (or `cutoff_role="recommended_max"` if we explicitly treat it as an upper recommended bound)

- **VASP `ENMIN`**  
  maps to: `cutoff_kind="wavefunction"`, `cutoff_role="recommended_min"`

- **QE `ecutwfc`**  
  maps to: `cutoff_kind="wavefunction"`, `cutoff_role="used"` (runtime input)

- **QE `ecutrho`**  
  maps to: `cutoff_kind="charge_density"`, `cutoff_role="used"` (runtime input)

### with this kind of skeleton, i am able to do the following as a user:

- “Show me PAW pseudopotentials with **recommended wavefunction** cutoff < X”
- “Show me entries where **used wavefunction** cutoff < **recommended**”
- “Show me pseudopotentials with **recommended_min**/**recommended_max** ranges and filter by max”


@ndaelman-hu @JFRudzinski

ndaelman-hu commented:
> To keep the schema code-agnostic while preserving queryability, we can model cutoffs as typed physical concepts rather than code-native labels like ENMAX or ecutwfc.

I like the general layout. This is basically a more grounded version of the OG `PhysicalProperty` and a pattern that has been applied in a few other places of the schema too.

I just have some questions / feedback regarding the `MEnum` values:

- `cutoff_kind`: would change "other" to "undecided" or "unknown", as it might be very very hard to establish for each code.
- `cutoff_role`:
  - would remove "used", as this is already extracted to `PlaneWaveBasisSet`. Or I define a more general `Cutoff` section. 
  - what is the distinction between "standard, stringent", "hard", and "soft"? I guess the first two have to do with the precision settings. Unfortunately, precision terminology is very code-dependent. We'd need a special research effort to standardize that. Hard and soft instead refer to kinds of pseudopotentials.

EBB2675 commented:
On `cutoff_kind`:
Yes, I agree that "other" is too vague. Renaming it to "unknown" (or "undecided") makes more sense semantically, since the ambiguity usually comes from lack of information rather than a genuinely different physical meaning :+1: 

On `cutoff_role = "used"`:
Agreed. If the actually used cutoff is already captured via `PlaneWaveBasisSet`, then it should not be duplicated here. In that case, `PPCutoff` should strictly represent pseudopotential-provided or library-provided guidance, not runtime settings :+1: please make sure to make this clear in the main docstring

About the MEnums:
I only included standard/stringent and hard/soft as examples to show how extendable the pattern is for the future, not as a strong claim that we already have a cross-code standard for these categories, so feel free to ditch them :+1:

ndaelman-hu committed:
Add `PPCutoff` as a specialized section

ndaelman-hu committed:
Polish descriptions and enum values in PPCutoff

ndaelman-hu committed:
Fix Pseudopotential SectionProxy resolution with full module path

ndaelman-hu committed:
Apply ruff

ndaelman-hu committed:
Merge branch 'develop' into add_pseudo_potential

ndaelman-hu commented:
@EBB2675 @JFRudzinski I applied Esma's suggestions. I just didn't get around to testing the nested queries yet.
Given the current timeframe and our aim to focus PR scope, I would open this as a follow-up issue.

Feel free to review, else I merge after lunch.

EBB2675 commented:
`PPCutoff` looks great

a quick question: how can i refer to the used cutoffs for different properties: wf, charge density? What does the `PlaneWaveBasisSet` record? is it only for the wavefunction cutoff?

ndaelman-hu commented:
> `PPCutoff` looks great
> 
> a quick question: how can i refer to the used cutoffs for different properties: wf, charge density? What does the `PlaneWaveBasisSet` record? is it only for the wavefunction cutoff?

Yeah, after pushing, I realized that I hadn't updated the basis set yet. Had been too engrossed in the query-side of things.
Will get back to you on this today.

ndaelman-hu commented:
> > `PPCutoff` looks great
> > a quick question: how can i refer to the used cutoffs for different properties: wf, charge density? What does the `PlaneWaveBasisSet` record? is it only for the wavefunction cutoff?
> 
> Yeah, after pushing, I realized that I hadn't updated the basis set yet. Had been too engrossed in the query-side of things. Will get back to you on this today.

I checked, but the distinction between the electronic structure property is supposed to be made at a higher level, in `BasisSetContainer`. There is a note there stating:

>  TODO: add reference to `electronic_structure`,
>  specifying to which electronic structure representation the basis set is applied
>  e.g. wavefunction, density, grids for subroutines, etc.

Given the current changes to the schema, this might be re-evaluated.
This means evaluating the whole basis set structure, which is beyond the scope of this PR.

All comments have been addressed, I'm merging this.
